### PR TITLE
Allowing to use a factory instead of manually instantiate the Jobs

### PR DESCRIPTION
**Pull request by serroba**
In order to make the Jobs injectable, I want to allow to delegate the responsibility of create new jobs to an independent factory. In that way, if I want to use Pimple, I can do so using a factory that will encapsulate access to Pimple to retrieve instances of `Resque_JobInterface`

I kept it as PSR-0 and I have added a couple of tests

### CODE REVIEW

**Comment by serroba** [comment]
ping @chrisboulton @philipmuir

File: test/Resque/Tests/JobTest.php [hunk 1]
```diff
@@ -362,4 +362,44 @@ public function testDequeueItemWithiWrongArg()
...
+        $testJob = $this->getMock(Resque_JobInterface::class);
+        $factory->expects(self::once())->method('create')->will($this->returnValue($testJob));
+        $instance = $job->getInstance();
+        $this->assertInstanceOf(Resque_JobInterface::class, $instance);
+    }
+
+    public function testDoNotUseFactoryToGetInstance()
+    {
+        $payload = array(
+            'class' => Some_Job_Class::class,
```
Comment (serroba) [inline test/Resque/Tests/JobTest.php]: oh... I see... If we need to keep supporting PHP 5.3/5.4 I can't make use of this... :(

What's your advice? I can easily replace with String. Not ideal, but... Whatever you prefer @chrisboulton

**Comment by danhunsaker** [comment]
I'm not convinced this will even work.  Recall that job classes are not instantiated (`->get_instance()`) by the same process that creates them (`::create()`) - the worker would have to set the factory for *every job* it pulls from the queue, and would either need a way to determine which jobs need it and which don't, or would not be capable of performing non-factory jobs.  And it would have to do this between when the job is pulled off the stack, and when its `->perform()` method is fired.  Which means you're either your changes are incomplete, here, or you're going to need to use the event hooks (`afterFork` is best, since it operate's in the child fork's process space, but `beforeFork` also fires in this interval) to fill out the details.  (I'll note that the event hooks are actually designed for use in just this manner, though I also wouldn't be against a [working] mechanism for this kind of work being built into the engine itself, either.)

The tests are a bit misleading in that they don't really show this split in the execution environment.  That's why I wrote the `HOWITWORKS` document (which I need to update, again, but which is still sound on the overall flow) - to give a better idea of what runs where, and which things you can expect to have access to at which times.

For myself, the method I'd use (if I wasn't using the event hooks) is to have a job class that doesn't do much by itself, but instead plugs its arguments into the factory I intend to use, and then passes control to the resulting object (if any) to do the actual work.  That gives everything you're looking for in one class, which is pretty nice to have, I think.

TL;DR - I like the idea, but I don't think this implementation is complete, nor is it necessarily the best approach, given the handful of alternatives already available, and the split execution environment inherent in any queues/workers system.

**Comment by chrisboulton** [comment]
Hey @danhunsaker - thanks for weighing in on this.. I was kinda hoping other people would. Full disclosure, @serroba works here with us (hence the funky BIG reference in the title of this PR).

I haven't had a moment to look over and really think about these changes yet - I was kinda hoping to do that over the next couple of days when things settle down and I can sort out the state of this project.

One thing that jumped out to me in your comment was this:

> For myself, the method I'd use (if I wasn't using the event hooks) is to have a job class that doesn't do much by itself, but instead plugs its arguments into the factory I intend to use, and then passes control to the resulting object (if any) to do the actual work.

This is actually what I was thinking before you even commented. There's an element of self-containedness that Resque promotes with jobs (hence why in the Ruby implementation it's kind of similar in that `Class::perform(args)` is called for every job). I think this leans toward what you're describing (at least in my mind) of having simple `perform` methods that basically just delegate work elsewhere by either instantiating or calling into some factory with the appropriate args/services. I think a trap we (and others) have fallen into is including a lot of their business logic in `perform` methods.

(Side note: this is making me wish more and more I'd kept parity with the Ruby implementation and had `perform` as a class method instead of instantiating an object.. It frustrates me right now that php-resque makes a lot of assumptions about how to instantiate something in your application)

Appreciate the feedback.

**Comment by chrisboulton** [comment]
Oh one other thing:

> Which means you're either your changes are incomplete, here, or you're going to need to use the event hooks (afterFork is best, since it operate's in the child fork's process space, but beforeFork also fires in this interval) to fill out the details

I _believe_ that's what @serroba planned to do in this case. We have a somewhat unique environment we run Resque in so we're pretty well already hooked in around these points.

**Comment by serroba** [comment]
I will wait to know what @chrisboulton has to said, but you are correct @danhunsaker on some of your assumptions.

The fact that the factory would have to be set on _every job_. But that is decided in the `afterFork` process and the factory is set only if the client decide to use a factory instead of the canonical mechanism.

I guess I can write test that reflect better the nature of how this works.

Regarding:
> "For myself, the method I'd use (if I wasn't using the event hooks) is to have a job class that doesn't do much by itself, but instead plugs its arguments into the factory I intend to use, and then passes control to the resulting object (if any) to do the actual work. That gives everything you're looking for in one class, which is pretty nice to have, I think"

That would be ideal, but the use case I need to cover has a very-hard-to-manage set of dependencies. So I don't have the luxury to be able to do that.

The whole intention of this PR is to add an _optional_ step of indirection to delegate control of instantiation of Jobs outside the direct control of the library. At this moment all the instantiation happens inside. I could of course, via inheritance, accomplish a similar idea, so instead to modify it in the library I do it via enhancing the library locally, but the idea to have a generic factory that anyone can decide to inject, seems to me a good idea.

**Comment by Benoit-amo** [comment]
Good PR :+1:

**Comment by danhunsaker** [comment]
@chrisboulton - Hrm.  The BIG prefix is on one of the commits, but not the PR itself.  I'll keep an eye out for it in the future, though.  :smile: 

@serroba - As I said, I'm not against support for factories in the project proper at all.  I was led a little astray by some assumptions i made on how you intended to use it - we're actually thinking about the same thing, here.  That said, if we want to support factories directly in the engine, here, I think we might want to approach it just a bit differently.

This PR already adds interfaces to implement depending on which type of job you're looking to create.  `Resque_JobInterface` for regular, non-factory jobs, and `Resque_Job_FactoryInterface` for factory-produced jobs.  Without quibbling on naming choices (names are *hard*), it's sensible to assume that the workers would check the interfaces for whether the current job class implements the Factory variant, and then grabs the factory from another source than a hook.  Say, for example, the developer is given the option to set the factory as a *worker* property, when the worker spins up.  The worker hands that factory to `Resque_Job`, and instantiation continues on its merry way.  No factory variant interface implementation? No worries - on you go.

(Of course, if we *really* want to be flexible, we *could* even include an option to pass multiple factories to the worker on startup, and then have the factory variant jobs provide a way to select which one to build the job class with.  But that's possibly a bit much.)

**Comment by danhunsaker** [comment]
@chrisboulton - Oh, and one other thing.  (Not terribly important, but I had to say it...)

In PHP we call them "static" methods, not "class" methods.  :grinning:

File: lib/Resque/Job.php [hunk 2]
```diff
@@ -182,7 +182,7 @@ public function getInstance()
 		}
 
         if ($this->jobFactory !== null) {
-            $this->instance = $this->jobFactory->create();
+            $this->instance = $this->jobFactory->create($this->payload['class']);
```
Comment (danhunsaker) [inline lib/Resque/Job.php]: Oops!  Even I missed this one...

**Comment by chrisboulton** [comment]
I think if this is to come down, it should be refactored to always work like this - job instantiation should always be performed via some kind of factory that you can override. There should be a default factory which is able to return a job instance that's configured and ready to go.

The factory interface should simply take the payload, and possibly the arguments or whatever else it needs, and return an instance of the job ready to call perform on. I think the logic for class loading, and existence check for the perform method should move there.

```
<?php
class Factory
{
  public function create($class, array $args, $queue)
  {
    if (!class_exists($class) {
      // ...
    }

    if (!method_exists($class, 'perform')) {
      // ...
    }

    $instance = new $class;
    $instance->args = $args;
    $instance->queue = $queue;
    return $instance;
  }
}
```

This leads to better portability for everyone.

**Comment by serroba** [comment]
Sure man. That's trivial to change. I'll update the PR

**Comment by danhunsaker** [comment]
To ensure the test runs the same way on every system regardless of its TZ settings.  Forcing users to switch to UTC in their INI every time they want to run a test...  Not best practice.

If it wasn't a test, it'd be bad practice.  It's a test, so it's actually preferred.  You want as predictable an environment as possible for testing, and keeping that in the test themselves ensures you get it.

**Comment by serroba** [comment]
What about now @chrisboulton ?

**Comment by serroba** [comment]
@philipmuir pointed to me that I was using spaces instead of tabs... Now is using tabs

**Comment by chrisboulton** [comment]
@serroba, mostly looks good - thanks ✨ 

For me, I was thinking that instead of having a code branch/conditional that sometimes calls into the factory to create a job, we should always do that and provide Resque a built in factory that's used if you haven't overloaded with a custom one via `setJobFactory`. To me, it seems cleaner to have one code path of creating job instances, vs two - I can see me in the future somehow screwing up and updating one code path and not the other.

WRT `Resque_Job` implementing the interface, I don't think we need that. People's jobs tend to be standalone classes, I don't know if anyone really extends their jobs from `Resque_Job`

File: lib/Resque/Job/FactoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+interface Resque_Job_FactoryInterface
+{
+	/**
+	 * @param $className
+	 * @param array $args
+	 * @param $queue
+	 * @return Resque_JobInterface
+	 */
+	public function create($className, $args, $queue);
```
Comment (serroba) [inline lib/Resque/Job/FactoryInterface.php]: @chrisboulton I had to remove the type hint here. As much as the tests are not sending an array here all the time (`null` !== `[]`) I imagine clients would be doing the same.

**Comment by Benoit-amo** [comment]
> WRT Resque_Job implementing the interface, I don't think we need that. People's jobs tend to be standalone classes, I don't know if anyone really extends their jobs from Resque_Job

@chrisboulton In my opinion, it's usefull. PHP Resqeue is a public library, with its own public API. Interfaces are a big part of the API. I my case, programming to interfaces is not an option. In my job, I'm used to abstract the external tools, like this. We are also using Zend JobQueue, but for smaller projects we use PHP Resque, so we have our own layer to abstract the queue.

**Comment by chrisboulton** [comment]
@serroba - looks good to me! 👍
